plugin/pkg/scheduler/core/scheduling_queue.go

```diff
@@ -15,47 +15,679 @@ limitations under the License.
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	if p.unschedulableQ.Get(pod) != nil {
+		return nil
+	}
+	if _, exists, _ := p.activeQ.Get(pod); exists {
+		return nil
+	}
+	err := p.activeQ.Add(pod)
+	if err != nil {
+		glog.Errorf("Error adding pod %v to the scheduling queue: %v", pod.Name, err)
+	} else {
+		p.cond.Broadcast()
+	}
+	return err
+}
+
+// AddUnschedulable adds a pod to the unschedulable queue.
+func (p *PriorityQueue) AddUnschedulable(pod *v1.Pod) error {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	if p.unschedulableQ.Get(pod) != nil {
+		return fmt.Errorf("pod is already present in unschedulableQ")
+	}
+	if _, exists, _ := p.activeQ.Get(pod); exists {
+		return fmt.Errorf("pod is already present in the activeQ")
+	}
+	if p.receivedMoveRequest {
+		return p.activeQ.Add(pod)
+	}
+	p.unschedulableQ.Add(pod)
+	return nil
+}
+
+// Pop removes the head of the active queue and returns it.
+func (p *PriorityQueue) Pop() (*v1.Pod, error) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	for len(p.activeQ.data.queue) == 0 {
+		p.cond.Wait()
+	}
+	obj, err := p.activeQ.Pop()
+	if err != nil {
+		return nil, err
+	}
+	p.receivedMoveRequest = false
+	return obj.(*v1.Pod), err
+}
+
+// isPodUpdated checks if the pod is updated in a way that it may have become
+// schedulable. It drops status of the pod and compares it with old version.
+func isPodUpdated(oldPod, newPod *v1.Pod) bool {
+	strip := func(pod *v1.Pod) *v1.Pod {
+		p := pod.DeepCopy()
+		p.ResourceVersion = ""
+		p.Status = v1.PodStatus{}
+		return p
+	}
+	return !reflect.DeepEqual(strip(oldPod), strip(newPod))
+}
+
+// Update updates a pod in the active queue if present. Otherwise, it removes
+// the item from the unschedulable queue and adds the updated one to the active
+// queue.
+func (p *PriorityQueue) Update(pod *v1.Pod) error {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	// If the pod is already in the active queue, just update it there.
+	if _, exists, _ := p.activeQ.Get(pod); exists {
+		err := p.activeQ.Update(pod)
+		if err == nil {
+			p.cond.Broadcast()
+		}
+		return err
+	}
+	// If the pod is in the unschedulable queue, updating it may make it schedulable.
+	if oldPod := p.unschedulableQ.Get(pod); oldPod != nil {
+		if isPodUpdated(oldPod, pod) {
+			p.unschedulableQ.Delete(pod)
+			err := p.activeQ.Add(pod)
+			if err == nil {
+				p.cond.Broadcast()
+			}
+			return err
+		} else {
+			p.unschedulableQ.Update(pod)
+			return nil
+		}
+	}
+	// If pod is not in any of the two queue, we put it in the active queue.
+	err := p.activeQ.Add(pod)
+	if err == nil {
+		p.cond.Broadcast()
+	}
+	return err
+}
+
+// Delete deletes the item from either of the two queues.
+func (p *PriorityQueue) Delete(pod *v1.Pod) error {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	if _, exists, _ := p.activeQ.Get(pod); exists {
+		return p.activeQ.Delete(pod)
+	}
+	p.unschedulableQ.Delete(pod)
+	return nil
+}
+
+// AssignedPodAdded is called when a bound pod is added. Creation of this pod
```

Comment:
Is this trying to say "AssignedPodAdded is called when a pod is bound to a node. Scheduling the pod may make pending pods with matching pod affinity terms schedulable"

If so, that would be clearer.

Reply:
What you said is a more probable scenario, but this event is also triggered when a pod whose node name is set is created.